### PR TITLE
utxo management and selection

### PR DESCRIPTION
**Pull request by bflood366**

### CODE REVIEW

**Review by antonyshchenko — APPROVED** [review]
Looks good!

File: backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/BitcoinClient.kt [hunk 1]
```diff
@@ -30,6 +30,30 @@ object BitcoinClient : JsonRpcClientBase(
...
+    fun mine(nBlocks: Int): List<String> {
+        return getValue(
+            BitcoinRpcRequest(
+                "generatetoaddress",
+                BitcoinRpcParams(listOf(nBlocks, bitcoinConfig.faucetAddress!!)),
+            ),
+        )
+    }
+
+    fun sendToAddress(address: String, amount: String): String {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/BitcoinClient.kt:42]: changed to a BigInteger and converted it internally to the string the RPC expects

[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/BitcoinClient.kt:42]: I changed this to a TxHash in a subsequent PR

File: backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/BitcoinClient.kt [hunk 2]
```diff
@@ -67,13 +91,13 @@ object BitcoinClient : JsonRpcClientBase(
...
 
-    open fun getRawTransaction(txId: String): BitcoinRpc.Transaction {
+    fun getRawTransaction(txId: String): BitcoinRpc.Transaction {
         return getValue(
             BitcoinRpcRequest(
                 "getrawtransaction",
                 BitcoinRpcParams(listOf(txId, true)),
             ),
-            false,
+            true,
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/BitcoinClient.kt:100]: whether we log the response body or not

File: backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/MempoolSpaceClient.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    data class MempoolApiFailure(
+        val error: String,
+    )
+}
+
+object MempoolSpaceClient {
+
+    val apiServerRootUrl = System.getenv("MEMPOOL_SPACE_API_URL") ?: "http://localhost:1080/api"
+    val httpClient = OkHttpClient.Builder().build()
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/MempoolSpaceClient.kt:63]: gonna leave as is for now and will add a ticket.

File: backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    fun selectUtxos(address: String, amount: BigInteger, fee: BigInteger): List<UnspentUtxo> {
+        return inputsSelector.selectInputs(
+            amount,
+            getUnspentUxtos(address),
+            fee,
+        )
+    }
+
+    fun getUnspentUxtos(address: String): List<UnspentUtxo> {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt]: changed the method name to refreshUnspectUtxos

File: backend/src/main/kotlin/xyz/funkybit/core/utils/bitcoin/BitcoinInputsSelector.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
+package xyz.funkybit.core.utils.bitcoin
+
+import xyz.funkybit.core.model.db.UnspentUtxo
+import java.math.BigInteger
+
+private typealias InputShuffleFn = (List<UnspentUtxo>) -> List<UnspentUtxo>
+
+class BitcoinInputsSelector(
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/utils/bitcoin/BitcoinInputsSelector.kt:8]: some scenarios are covered by the `UtxoSelectionTest` but will add a ticket to bring more of the unit tests from prime in

File: backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/MempoolSpaceClient.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        encodeDefaults = true
+        isLenient = true
+        explicitNulls = false
+    }
+    val logger = KotlinLogging.logger {}
+
+    private fun execute(request: Request): Response =
+        httpClient.newCall(request).execute()
+
+    private fun tryGetTransactions(walletAddress: String, afterTxId: String?): Either<MempoolSpaceApi.MempoolApiFailure, List<MempoolSpaceApi.Transaction>> {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/blockchain/bitcoin/MempoolSpaceClient.kt]: changed

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/BitcoinWalletState.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+data class UnspentUtxo(
+    val utxoId: String,
+    val amount: BigIntegerJson,
+    val blockHeight: Long?,
+)
+
+object BitcoinWalletStateTable : GUIDTable<BitcoinWalletStateId>("bitcoin_wallet_state", ::BitcoinWalletStateId) {
+    val address = varchar("address", 10485760).uniqueIndex()
+    val createdAt = timestamp("created_at")
+    val createdBy = varchar("created_by", 10485760)
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/model/db/BitcoinWalletState.kt]: removed - should probaby do elsewhere too.

**Pull request merged**